### ISSUE TITLE
conversion_plus creating non output carrier? related to the carrier_out name

### ISSUE THREAD

**Issue by arnaud-leroy** [calliope-project/calliope#60]
Hello,

I am testing conversion_plus technology after seeing some strange behavior while using it.

I created a model with a power demand, a power supply and a conversion plus technology consuming power to release another carrier.

```
techs:
    supply_power:
        name: 'Electricity import'
        parent: supply
        carrier: power
        constraints:
            r: inf
            e_cap.max: 100
        costs:
            monetary:
                e_cap: 1
                om_fuel: 50

    conv_tech:
        name: 'conversion tech consuming power'
        parent: conversion_plus
        carrier_in: power    
        carrier_out:   
            name_carrier_out: 0.01
        primary_carrier: name_carrier_out

    demand_power:
        name: 'Electrical demand'
        parent: demand
        carrier: power

    unmet_demand_power:
        name: 'Unmet electrical demand'
        parent: unmet_demand
        carrier: power

```

In fact the name of the carrier_out has an influence on the result of the model: 
- In general, the power supply will provide the power for the demand. 
![image](https://user-images.githubusercontent.com/29625919/28862535-a5e39c0a-7765-11e7-8f8e-41da9ef14e0c.png)

- But if the name of the carrier out contains 'power', like for example `        carrier_out:   
           power_out_conversion: 0.01`, or `        carrier_out:   
            test_power_out_conversion: 0.01` the conversion_plus technology ends up feeding the power demand
![image](https://user-images.githubusercontent.com/29625919/28862557-baadcd2c-7765-11e7-8f83-548b51367e79.png)

- If I just change the name of the carrier_out to:  `        carrier_out:   
            test_pow_er_out_conversion: 0.01` the supply is once again feeding the power demand
![image](https://user-images.githubusercontent.com/29625919/28862578-cc3cd362-7765-11e7-81b0-80c464d2b181.png)

I also changed the primary_carrier every time to match the carrier_out name.

**Comment by arnaud-leroy** [comment]
Another example. This large_Gas_Boiler is providing heat instead of district_heat:
```
  large_Gas_Boiler:
    carrier_in: gas
    carrier_out:
      district_heat: 0.9
    constraints:
      e_cap.max: .inf
    costs:
      monetary:
        e_cap: 6100000.0
        om_fixed: 109800.0
    depreciation:
      plant_life: 21.0
    name: Gas Boiler
    parent: conversion_plus
    primary_carrier: district_heat

```

**Comment by brynpickering** [comment]
Thanks for this Arnaud. It was being caused by a in constraint generation:

`carriers_out = model.get_carrier(y, 'out', all_carriers=True)
            if c not in carriers_out:
                return c_prod == 0
            if c in carriers_out and model._locations.at[x, y] == 0:
                return c_prod == 0`
In the examples you give, with only one carrier out, `model.get_carrier` returns a string e.g. `district_heat`. So the carrier `heat` would be flagged as being "in" `district_heat`. If there were multiple output carriers, e.g. `district_heat` and `power`, `model.get_carrier` would return a tuple (`power`, `district_heat`), and `heat` would not be flagged as "in" this.

Conversion plus has quite a few quirks, I'll see if this can be resolved easily and then aim to fix it. If it causes issues downstream then it could take me a bit longer to sort.

**Comment by brynpickering** [comment]
Fixed in commit cc9536c
